feat: scale and map body

jonahrb opened the pull request:
## Description
Ability to scale or map a body

## Issue linked
#1036 

## Checklist
- [x] I have tested my changes locally.
- [x] I have added necessary documentation or updated existing documentation.
- [x] I have followed the coding style guidelines of this project.
- [x] I have added appropriate unit tests.
- [x] I have reviewed my changes before submitting this pull request.
- [x] I have linked the issue or issues that are solved to the PR if any.
- [x] I have assigned this PR to myself.
- [x] I have added the minimum version decorator to any new backend method implemented.
- [x] I have made sure that the title of my PR follows [Conventional commits style](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. ``feat: extrude circle to cylinder``)

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -358,6 +362,42 @@ def rotate(
         return
 
+    @abstractmethod
+    def scale(
+        self,
+        value: Real,
+    ) -> None:
```
I'd single line them if possible
```suggestion
    def scale(self, value: Real) -> None:
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -358,6 +362,42 @@ def rotate(
+        Parameters
+        ----------
+        value: Real
+            Value to scale the body by.
+        Returns
+        -------
+        None
```
If nothing is returned, no need to add the return section

```suggestion
            Value to scale the body by.
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -358,6 +362,42 @@ def rotate(
+        """
+        return
+
+    @abstractmethod
+    def mirror(
+        self,
+        frame: Frame,
+    ) -> None:
```
```suggestion
    @abstractmethod
    def mirror(self, frame: Frame) -> None:
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -358,6 +362,42 @@ def rotate(
+        Parameters
+        ----------
+        frame: Frame
+            Structure defining the mirroring/mapping of the body.
+        Returns
+        -------
+        None
```
```suggestion
            Structure defining the mirroring/mapping of the body.
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -818,6 +858,25 @@ def rotate(
+    @protect_grpc
+    @check_input_types
+    @reset_tessellation_cache
+    @min_backend_version(24, 1, 0)
```
Oh! It was already available for all backends?

jonahrb replied:
@RobPasMue No, but i was able to retroactively add it, so it is technically available with 24.1 but they would need an updated service :)

RobPasMue replied:
In that case... let's avoid "retroactive" behavior. I talked about this with @b-matteo. Your service is the one that defines the backend version number so you need at least a 24R2 service to do this. Meaning that we need to add the decorators as 24, 2, 0

jonahrb replied:
@RobPasMue makes sense, on it

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -818,6 +858,25 @@ def rotate(
+    @check_input_types
+    @reset_tessellation_cache
+    @min_backend_version(24, 1, 0)
+    def mirror(
+        self,
+        frame: Frame,
+    ) -> None:  # noqa: D102
```
```suggestion
    def mirror(self, frame: Frame) -> None:  # noqa: D102
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -818,6 +858,25 @@ def rotate(
+    @protect_grpc
+    @check_input_types
+    @reset_tessellation_cache
+    @min_backend_version(24, 1, 0)
```
If that is the case (it was already available in 24R1) no need to add the decorator. We assume the first version is 24R1.
```suggestion
```

RobPasMue commented on src/ansys/geometry/core/designer/body.py:
```diff
@@ -818,6 +858,25 @@ def rotate(
+    @protect_grpc
+    @check_input_types
+    @reset_tessellation_cache
+    @min_backend_version(24, 1, 0)
```
Same as before here
```suggestion
```

RobPasMue reviewed (approved):
Looking good @jonahrb

jonahrb commented:
@RobPasMue FYI i renamed "mirror" to "map" since that's a more accurate description of the method. I also added unit tests, see `test_body_mapping()` for a couple examples

PipKat commented on tests/integration/test_design.py:
```diff
@@ -1927,3 +1927,111 @@ def test_get_collision(modeler: Modeler):
+
+
+def test_body_scale(modeler: Modeler):
+    """Test for verifying the correct scaling of a ``Body``."""
```
```suggestion
    """Verify the correct scaling of a body."""
```

RobPasMue replied:
a body works @PipKat - thanks for pointing it out nonetheless :)

PipKat commented on tests/integration/test_design.py:
```diff
@@ -1927,3 +1927,111 @@ def test_get_collision(modeler: Modeler):
+
+
+def test_body_mapping(modeler: Modeler):
+    """Test for verifying the correct mapping of a ``Body``."""
```
```suggestion
    """Verify the correct mapping of a body."""
```

RobPasMue reviewed (dismissed):
Tests are great, implementation is great. Thanks a lot @jonahrb - this is going to be extremely helpful. I would only suggest adding some examples in our documentation on how to do these operations. But it can be done in another PR. I will create an issue and assign it to you.

dastan-ansys reviewed (dismissed):
looks good! I am not big on documentation personally, but I feel like it would be a good idea to add a comment that describes the concrete implementation of scale() and map() or links them back the comment on their abstract counterparts.
